src/RoleObjects/GridData.js

```diff
@@ -65,71 +82,345 @@ export default class GridData extends TableData {
-      let colIndex = this.getColIndex(target);
-      const currentRowData = rowArr[rowIndex];
-      const isExpandable = currentRowData.expandedArrow && currentRowData.expandedArrow.visible;
-      const isCollapsable = currentRowData.collapsedArrow && currentRowData.collapsedArrow.visible;
-      const rowClick = (e) => {
-        const evt = new createjs.Event('keyboardClick', false, e.cancelable);
-        const skipPreventDefault = currentRowData.dispatchEvent(evt);
-        if (!skipPreventDefault) {
-          e.preventDefault();
-        }
-        e.stopPropagation();
-      };
-      switch (event.keyCode) {
-        case KeyCodes.up:
-          rowIndex--;
-          break;
-        case KeyCodes.down:
-          rowIndex++;
-          break;
-        case KeyCodes.left:
-          isExpandable ? rowClick(event) : colIndex--;
-          break;
-        case KeyCodes.right:
-          isCollapsable ? rowClick(event) : colIndex++;
-          break;
-        case KeyCodes.home:
-          colIndex = 0;
-          break;
-        case KeyCodes.end:
-          colIndex = currentRowData.cellCount - 1;
-          break;
-        default:
-          break;
+  /**
+   * @inheritdoc
+   */
+  _onKeyDown(evt) {
+    if (this.enableKeyEvents) {
+      super._onKeyDown(evt);
+      if (evt.defaultPrevented) {
+        return;
       }
+    }
+
+    const keyToUpdateMap = {
+      [KeyCodes.up]: '_updateTargetDataUp',
+      [KeyCodes.down]: '_updateTargetDataDown',
+      [KeyCodes.left]: '_updateTargetDataLeft',
+      [KeyCodes.right]: '_updateTargetDataRight',
+      [KeyCodes.home]: '_updateTargetDataHome',
+      [KeyCodes.end]: '_updateTargetDataEnd',
+    };
+
+    const updateFuncName = keyToUpdateMap[evt.keyCode];
+    if (updateFuncName) {
+      const targetData = this._focusableElemToTargetData(evt.target);
+      const attemptFocusUpdate = this[updateFuncName](targetData);
+
+      if (attemptFocusUpdate) {
+        const nextTarget = this._findNextTarget(targetData);
+        this._focusToNextTarget(nextTarget, targetData, evt);
+      }
+    }
+  }
+
+  /**
+   * Handles updating the target data when the up arrow key is pressed
+   * @access protected
+   *
+   * @param {!TargetData} targetData - target data to update/mutate for which
+   * DisplayObject has focus vs which should get focus as a result of the
+   * keystroke
+   * @returns {boolean} true if focus should potentially be moved,
+   * false if potentially moving focus should be skipped (e.g. the treegrid
+   * emitting an event to handle expanding/collapsing a row instead of moving
+   * focus)
+   */
+  _updateTargetDataUp(targetData) {
+    targetData.rowIndex--;
+    if (targetData.rowIndex < 0 && targetData.sectionIndex > 0) {
+      // handles rolling into the prior table section
+      targetData.sectionIndex--;
+      targetData.rowIndex = this.children[targetData.sectionIndex].accessible.children.length - 1; // eslint-disable-line max-len
```

Comment:
max-len can be dealt with more properly in the future if/when prettier is integrated into this repo